pkg/leaderelection/leaderelection.go

```diff
@@ -47,18 +43,42 @@ type LeaderElector struct {
 // Run starts the leader election loop. Run will not return
 // before leader election loop is stopped by ctx or it has
 // stopped holding the leader lease
-func (le *LeaderElector) Run(ctx context.Context) {
-	defer func() {
-		le.config.Callbacks.OnStoppedLeading()
-	}()
+func (le *LeaderElector) Run(ctx context.Context, wg *sync.WaitGroup) {
+	defer wg.Done()
 
-	if !le.acquire(ctx) {
-		return // ctx signalled done
-	}
-	ctx, cancel := context.WithCancel(ctx)
-	defer cancel()
-	go le.config.Callbacks.OnStartedLeading(ctx)
-	le.renew(ctx)
+	leadCtx, leadCancel := context.WithCancel(ctx)
+
+	logger.Ctx(ctx).Info("attempting to acquire leader lease")
+	// outer wait loop runs when a instance has successfully acquired lease
+	wait.Until(func() {
+		retryCtx, retryCancel := context.WithCancel(ctx)
+		defer retryCancel()
+
+		// inner wait retry loop for faster retries if failed to acquire lease
+		acquired := false
+		wait.Until(func() {
+			acquired = le.tryAcquireOrRenew(retryCtx)
+
+			if !acquired {
+				logger.Ctx(retryCtx).Infow("failed to acquire lease, retrying...")
+				return
+			}
+
+			// if succeeded, we can break the inner loop by cancelling context and renew in outer loop
+			logger.Ctx(ctx).Info("successfully acquired lease")
+			retryCancel()
+		}, le.config.RetryPeriod, retryCtx.Done())
+
+		// OnStartedLeading
+		le.leaderID = le.nodeID
+		go le.config.Callbacks.OnStartedLeading(leadCtx)
+
+	}, le.config.RenewDeadline, ctx.Done())
+
+	// context returned done, release the lease
+	logger.Ctx(ctx).Info("context done, releasing lease")
+	leadCancel()
+	le.release(ctx)
```

Comment:
would need to wait for `OnStartedLeading` to successfully terminate before returning from this function

Reply:
release calls onStopLeading callback, that will ensure that we no longer perform leader tasks